charts/hub-gateway/plugins/credits.lua

```diff
@@ -0,0 +1,118 @@
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core = require("apisix.core")
+local http = require("resty.http")
+local json = require("apisix.core.json")
+
+local schema = {
+    type = "object",
+    properties = {
+        host = {
+            type = "string"
+        },
+        ssl_verify = {
+            type = "boolean",
+            default = true
+        },
+        timeout = {
+            type = "integer",
+            minimum = 1,
+            maximum = 60000,
+            default = 3000,
+            description = "timeout in milliseconds"
+        },
+        keepalive = {
+            type = "boolean",
+            default = true
+        },
+        keepalive_timeout = {
+            type = "integer",
+            minimum = 1000,
+            default = 60000
+        },
+        keepalive_pool = {
+            type = "integer",
+            minimum = 1,
+            default = 5
+        },
+        org_cookie_name = {
+            type = "string"
+        },
+    },
+    required = {"host"}
+}
+
+local _M = {
+    version = 0.1,
+    priority = 2,
+    name = "credits",
+    schema = schema
+}
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+function _M.access(conf, ctx)
+    local org_cookie_name = conf.org_cookie_name or "_hub_org"
+    local cookie_header = "cookie_" .. org_cookie_name
+    local cookie_value = ngx.var[cookie_header]
+    
+
+    -- return early if operation is query
+    local operation = ctx.var.graphql_operation
+    if not operation or operation == "query" then
+        return
+    end
+
+    local params = {
+        method = "GET",
+        keepalive = conf.keepalive,
+        ssl_verify = conf.ssl_verify
+    }
+
+    if conf.keepalive then
+        params.keepalive_timeout = conf.keepalive_timeout
+        params.keepalive_pool = conf.keepalive_pool
+    end
+
+    local endpoint = conf.host .. "/internal/organizations/" .. cookie_value
+    local httpc = http.new()
+    httpc:set_timeout(conf.timeout)
+    local res, err = httpc:request_uri(endpoint, params)
+
+    -- return internal server error if unable to contact credits service
+    if not res then
+        return 500, json.encode({ message = err })
+    end
+
+    local data, err = json.decode(res.body)
+    if not data then
+        return 500, json.encode({ message = err })
+    end
+
+    if not data.balance then
+        return 500, json.encode({ message = err })
+    end
+
+    if conf.expose_user_id then
+        core.request.set_header(ctx, "X-CREDIT-BALANCE", data.balance)
+        core.response.set_header(ctx, "X-CREDIT-BALANCE", data.balance)
```

Comment:
X-CREDIT-BALANCED repeated twice.

Reply:
We also need the X-Organization-ID. To compose the plugins may be a good candidate for a dedicated plugin to consilated the cookie or client app lookup logic.

Reply:
> X-CREDIT-BALANCED repeated twice.

Its `core.request` and `core.response`, so we also reply back the credit balance to the user through headers.

I think its cool to let them know their balance constantly if they are not using Hub through the UI.


> We also need the X-Organization-ID. To compose the plugins may be a good candidate for a dedicated plugin to consilated the cookie or client app lookup logic.

Yea, agree. I'll add a `organization` plugin in the PR to move `x-client-owner-id` and `_hub_org` cookie to `x-organization-id` header